docs/modules/ROOT/pages/high_level/security_group_host_design.adoc

```diff
@@ -0,0 +1,583 @@
+. add a new class for the main security group implementation, add default openflow rules in constructor and clean up the openflow rules in destructor. Note that a port can have "port security" enabled (default) or disabled. Therefore, any default openflow rules need to work on both.
+. for openflow rule manipulation, use ACA_OVS_Control::get_instance().add_flow/mod_flows/del_flows.
+. modify CMakeLists.txt under /src to include new .cpp files.
+. modify aca_comm_mgr.cpp - Aca_Comm_Manager::update_goal_state - add a new block for Aca_SG_State_Handler::get_instance().update_sg_states similar to update_dhcp_states.
+. update test/gtest/aca_tests.cpp to include a set of unit and functional test cases for security group implementation. We may need to create a docker container, use ovs-docker to connect its port to br-int. Then push down a crafted goal state to configure a new port with a set of security group rules. The last step is to send some traffic to confirm the security group rules are working. See DISABLED_2_ports_ROUTING_test_traffic_one_machine in the same file for reference.
+
+We will need to create/update the corresponding header files accordingly.
+
+=== Integration with Alcor Distributed Router
+
+Alcor Distribute Router was using table 0 to restore gateway mac address from the neighbor host DVR mac. 
+With addition of security group implementation, 
+we will move that to table 73 (ACCEPT_OR_INGRESS) since those connection is already accepted for ingress pipeline.
+
+
+== VxLAN-GPE for remote security group
+
+We can consider leveraging VxLAN-GPE to address the scale problem with security group, on rules that contain remote security group as the number of ports (and assoicated IPs) on the remote group can be huge. That will require local host to be aware of all the ports assoicated IPs to construct the corresponding openflow rules.
+
+While the current design still need to work with VxLAN only environment, we can leveraging VxLAN-GPE to embed remote security group IDs or markings into ingress packets. The local host can then apply ingress SG rule without needing to know all the ports IPs from the remote group beforehand.
+
+=== VxLAN-GPE packet with remote security group info
+
+[source,shell]
+------------------------------------------------------------
+Outer header -- Next header ------------ Inner packet
+VxLAN-GPE ----- NSH with remote SG ID -- IPv4/IPv6 original payload
+------------------------------------------------------------
+
+=== OVS egress marking
+
+OVS (controlled by ACA) will need to stamp all packets coming out for a port with the assoicated SG IDs (zero to five).
+
+Each SG ID is 36 characters string (36 bytes) or we can represent it using 16 bytes encoding. 5 of them will be 80 bytes plus adding 8 bytes NSH header will total 88 bytes increase of packet header. While VxLAN-GPE + NSH can embed all those information, but 88 bytes is too much data to put into header of every single packet. 
```

Comment:
Assuming that most majority of customer scenarios only use one SG, can we cut the header overhead to 16+5 = 21 bytes or 88 is not negotiable :-)

Reply:
that would cut it down to 16+8 bytes NSH header = 24 bytes. That's more negotiable :) @chenpiaoping @Gzure  - can we assume most customer scenarios only use one SG?

Reply:
Let us bring this question up in the weekly open-source meeting. If the assumption is confirmed, we should take the proposal more serious :-)